proposals/0021-vk-coop-matrix.md

```diff
@@ -0,0 +1,264 @@
+<!-- {% raw %} -->
+
+# Vulkan cooperative matrix extension
+
+*   Proposal: [0021](0021-vk-coop-matrix.md)
+*   Author(s): [Steven Perron](https://github.com/s-perron)
+*   Sponsor: TBD
+*   Status: **Accepted**
+*   Required Version: Vulkan 1.1
+*   PRs: [#6720](https://github.com/microsoft/DirectXShaderCompiler/pull/6720)
+
+## Introduction
+
+A new type, `vk::khr::CooperativeMatrix`, with functions to act on them will be
+added to allow HLSL shaders to expose
+[VK_KHR_cooperative_matrix](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VK_KHR_cooperative_matrix.html).
+
+## Motivation
+
+Users have no user-friendly way to write Vulkan shaders that use cooperative
+matrices. This is a useful feature that users would like to use.
+
+## High-level description
+
+The solution to this problem is to add a new class,
+`vk::khr::CooperativeMatrix`, which will be defined in a header file
+"vk/khr/cooperative_matrix.h". This class will create an object with Spir-V type
+`OpTypeCooperativeMatrixKHR`. Functions are added that will expose the Spir-V
+the operations that take cooperative matrices as operands. All functions are
+defined to match the corresponding operations in the
+(SPV_KHR_cooperative_matrix)[https://htmlpreview.github.io/?https://github.com/KhronosGroup/SPIRV-Registry/blob/main/extensions/KHR/SPV_KHR_cooperative_matrix.html]
+extension.
+
+To help implement `vk::khr::CooperativeMatrix`, we will also introduce a utility
+class `vk::util::ArithmeticSelector`. This class will be defined in a header
+file "vk/util/arithmetic_selector.h". It can be used to generate inline Spir-V
+arithmetic instructions with the correct opcode for the type.
+
+The interface for `vk::khr::CooperativeMatrix` uses enums that are defined in
+the Spir-V specification. The required enums will be defined in "vk/spirv.h".
+
+## Detailed design
+
+### `vk::khr::CooperativeMatrix`
+
+The `CooperativeMatrix` class will defined in a header file as a wrapper around
+a `vk::SpirvType` that will expand to the appropriate SPIR-V type. This class
+will have the following interface.
+
+```c++
+
+// The base cooperative matrix class. The template arguments correspond to the
+// operands in the OpTypeCooperativeMatrixKHR instruction.
+template <typename ComponentType, Scope scope, uint rows, uint columns,
+CooperativeMatrixUse use>
+class CooperativeMatrix {
```

Comment:
There are some operations that the spir-v spec allows that are missing:

- [ ] Constructing a coop  matrix from a single scalar
- [ ] OpConvertFToU
- [ ] OpConvertFToS
- [ ] OpConvertSToF
- [ ] OpConvertUToF
- [ ] OpUConvert
- [ ] OpSConvert
- [ ] OpFConvert
- [ ] OpBitCast
- [ ] accessing individual elements (This will add access chains that spirv-opt will have to optimize).

Reply:
All of the operations have been added to the header file implementation in DXC. Still need to update this PR.